src/core/Akka/Event/EventStream.cs

```diff
@@ -89,9 +97,83 @@ public override bool Unsubscribe(IActorRef subscriber)
             {
                 Publish(new Debug(SimpleName(this), GetType(), "unsubscribing " + subscriber + " from all channels"));
             }
-
+            UnregisterIfNoMoreSubscribedChannels(subscriber);
             return base.Unsubscribe(subscriber);
         }
+
+        public void StartUnsubscriber(ActorSystem system)
+        {
+            EventStreamUnsubscribersProvider.Instance.Start(system, this, _debug);
+        }
+
+        public bool InitUnsubscriber(IActorRef unsubscriber, ActorSystem system)
+        {
+            if (system == null)
+            {
+                return false;
+            }
+            return _initiallySubscribedOrUnsubscriber.Match().With<Left<ISet<IActorRef>>>(v =>
+            {
+                if (_initiallySubscribedOrUnsubscriber.CompareAndSet(v, Either.Right(unsubscriber)))
+                {
+                    if (_debug)
+                    {
+                        Publish(new Debug(SimpleName(this), GetType(),
+                            $"initialized unsubscriber to: {unsubscriber}  registering {v.Value.Count} initial subscribers with it"));
+
+                    }
+                    v.Value.ForEach(RegisterWithUnsubscriber);
+
+
+                }
+                else
+                {
+                    InitUnsubscriber(unsubscriber, system);
+                }
+
+
+            }).With<Right<IActorRef>>(presentUnsubscriber =>
+            {
+                if (_debug)
+                {
+                    Publish(new Debug(SimpleName(this), GetType(),
+                        $"not using unsubscriber {unsubscriber}, because already initialized with {presentUnsubscriber}"));
+
+                }
+            }).WasHandled;
+        }
+
+        private void RegisterWithUnsubscriber(IActorRef subscriber)
+        {
+            _initiallySubscribedOrUnsubscriber.Match().With<Left<ISet<IActorRef>>>(v =>
+            {
+                if (_initiallySubscribedOrUnsubscriber.CompareAndSet(v,
+                    Either.Left<ISet<IActorRef>>(new HashSet<IActorRef>(v.Value.Union(new[] {subscriber})))))
```

Comment:
Immutable set is better choice here, we probably don't want to rebuild a whole `HashSet` on each change. Same concerns line 167.